opentelemetry-sdk/src/trace/span_processor_with_async_runtime.rs

```diff
@@ -6,18 +6,20 @@ use crate::trace::Span;
 use crate::trace::SpanProcessor;
 use crate::trace::{SpanData, SpanExporter};
 use futures_channel::oneshot;
-use futures_util::pin_mut;
 use futures_util::{
     future::{self, BoxFuture, Either},
-    select,
+    pin_mut, select,
     stream::{self, FusedStream, FuturesUnordered},
-    StreamExt as _,
+    FutureExt as _, StreamExt as _, TryFutureExt as _,
 };
 use opentelemetry::Context;
 use opentelemetry::{otel_debug, otel_error, otel_warn};
+use std::collections::VecDeque;
 use std::fmt;
+use std::future::Future;
 use std::sync::atomic::{AtomicUsize, Ordering};
 use std::sync::Arc;
+use tokio::sync::RwLock;
```

Comment:
We can't assume that the Tokio runtime is always available in this context. Instead, the [runtime abstraction](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-sdk/src/runtime.rs) is used to ensure flexibility and avoid locking the code to a specific async runtime.

Reply:
@lalitb Please read the following: https://docs.rs/tokio/latest/tokio/sync/index.html#runtime-compatibility

Reply:
Thanks for link. This makes sense. However, I was just thinking of the scenario where `experimental_async_runtime` and `rt-async-std` features are enabled for otel-sdk. I believe the compilation will fail in that case, as tokio being optional dependency. Or am I missing something.

Reply:
That makes me think - we don't have CI test for this :)

Reply:
@lalitb Makes sense. We can make `experimental_async_runtime` to depend on `tokio/sync` (it contains only Tokio sync primitives) to avoid importing some other libraries containing asynchronous RwLock. The result will be
```toml
//Cargo.toml
experimental_async_runtime = ["dep:tokio", "tokio/sync"]
rt-tokio = ["dep:tokio", "dep:tokio-stream", "tokio/rt", "tokio/time", "experimental_async_runtime"]
rt-tokio-current-thread = ["dep:tokio", "dep:tokio-stream", "tokio/rt", "tokio/time", "experimental_async_runtime"]
rt-async-std = ["dep:async-std", "experimental_async_runtime"]
```

Reply:
Looks good to me. Also, another option could be to use the [async_std::sync::RwLock](https://docs.rs/async-std/latest/async_std/sync/struct.RwLock.html) for rt-async-std feature.

Reply:
@lalitb

> Also, another option could be to use the [async_std::sync::RwLock](https://docs.rs/async-std/latest/async_std/sync/struct.RwLock.html)

This will bring additional `#[cfg]` attributes depending on runtime, which will increase the code complexity. As long Tokio's RwLock works for any runtime, consider using it.

Also, here's the demo of using `async-std`. Everything complies (as we discussed above), but it seems like `opentelemetry-otlp` crate relying on some Tokio stuff:
https://github.com/50U10FCA7/otel-oom/tree/debug-async-std

Reply:
> Everything complies (as we discussed above), but it seems like opentelemetry-otlp crate relying on some Tokio stuff (opentelemetry-otlp uses hyper-util which relies on Tokio runtime). Not part of this PR, but just mentioning.

Yes, this is expected for otlp/gRPC and otlp/http with hyper.